pulsar-broker/src/main/java/org/apache/pulsar/compaction/TwoPhaseCompactor.java

```diff
@@ -400,6 +400,16 @@ private static Pair<String, Integer> extractKeyAndSize(RawMessage m) {
         }
     }
 
+    protected List<ImmutableTriple<MessageId, String, Integer>> extractIdsAndKeysAndSizeFromBatch(RawMessage msg)
+            throws IOException {
+        return RawBatchConverter.extractIdsAndKeysAndSize(msg);
+    }
+
+    protected Optional<RawMessage> rebatchMessage(RawMessage msg, BiPredicate<String, MessageId> filter)
+            throws IOException {
+        return RawBatchConverter.rebatchMessage(msg, filter);
+    }
```

Comment:
This is easy to Kop extend `TwoPhaseCompactor` and overwriting them.

Reply:
It's not a good idea just for this motivation. We should not apply any guarantee to the protected methods. We don't have legal reasons to prevent others from removing them or just making them private.

It's not friendly for any refactoring. Others might change the method to return a nullable `RawMessage` or even not return a `RawMessage`.

Reply:
Let's keep it `protected` for now.